### ISSUE TITLE
Uri percent-encoding inconsistency between .NET Framework and UWP

### ISSUE THREAD

**Issue by linvi** [dotnet/corefx#6946]
I am not sure this is the best place to notify Microsoft about the following problem but here we go :

I have noticed an inconsistency of HttpClient.SendAsync between "Console" and W10 Universal App.
The inconsistency can be detected via Fiddler when executing a query with special characters.

Please note that I am using the `HttpClient` from a Portable Class Library.

If I user `HttpClient.SendAsync` the following query : https://api.twitter.com/1.1/search/tweets.json?q=%D0%9F%D1%80%D0%B8%D0%B2%D0%B5%D1%82%21&result_type=mixed&count=100, the query executed by .NET is different if it is run in UAP and if it is run in a Console application. (query = **Привет!**)

I have been able to identify this via Fiddler:

**Console Application :**

> GET https://api.twitter.com/1.1/search/tweets.json?q=%D0%9F%D1%80%D0%B8%D0%B2%D0%B5%D1%82!&result_type=mixed&count=100 HTTP/1.1
> 
> User-Agent: Tweetinvi/0.9.11.1
> 
> Cache-Control: no-cache
> 
> Authorization: OAuth oauth_consumer_key="jjMV4k3n9EswD9hlhRZqQCZrm",oauth_nonce="222222",oauth_signature_method="HMAC-SHA1",oauth_timestamp="1458086400",oauth_token="1693649419-BlEivyWIiOVrb22JjdzRipXWp4ltVdo4VLye1VW",oauth_version="1.0",oauth_signature="ODaUZNfxMbM7l0gtZ1GEFsjr%2BAA%3D"
> 
> Host: api.twitter.com
> 
> Connection: Keep-Alive

**Windows 10 Universal App :**

> GET https://api.twitter.com/1.1/search/tweets.json?q=%D0%9F%D1%80%D0%B8%D0%B2%D0%B5%D1%82%21&result_type=mixed&count=100 HTTP/1.1
> 
> Cache-Control: no-cache
> 
> User-Agent: Tweetinvi/0.9.11.1
> 
> Authorization: OAuth oauth_consumer_key="jjMV4k3n9EswD9hlhRZqQCZrm",oauth_nonce="222222",oauth_signature_method="HMAC-SHA1",oauth_timestamp="1458086400",oauth_token="1693649419-BlEivyWIiOVrb22JjdzRipXWp4ltVdo4VLye1VW",oauth_version="1.0",oauth_signature="ODaUZNfxMbM7l0gtZ1GEFsjr%2BAA%3D"
> 
> Host: api.twitter.com
> 
> Connection: Keep-Alive
> 
> Cookie: guest_id=v1%3A145704170375190067

If you take attention, you will notice that on the console, the end of the query parameter is **%82!** (with a question mark). 

On the other side the W10 query parameter finishes by **%82%21**. Here the question mark has kept its HTML escaping.

You can also notice that W10 is adding a cookie.

Is it a bug of the .NET Framework/HttpClient library?

I can assure that the query entered is the SAME in the `HttpClient.SendAsync` as both application uses the same binaries to create and execute the query (and I verified).

Here is my code for reference:

``` c#
    public class HttpClientWebHelper : IHttpClientWebHelper
    {
        public async Task<HttpResponseMessage> GetHttpResponse(ITwitterQuery twitterQuery, HttpContent httpContent = null, ITwitterClientHandler handler = null)
        {
            using (var client = GetHttpClient(twitterQuery, handler))
            {
                client.Timeout = twitterQuery.Timeout;

                var httpMethod = new HttpMethod(twitterQuery.HttpMethod.ToString());

                if (httpContent == null)
                {
                    return await client.SendAsync(new HttpRequestMessage(httpMethod, twitterQuery.QueryURL)).ConfigureAwait(false);
                }
                else
                {
                    if (httpMethod != HttpMethod.Post)
                    {
                        throw new ArgumentException("Cannot send HttpContent in a WebRequest that is not POST.");
                    }

                    return await client.PostAsync(twitterQuery.QueryURL, httpContent).ConfigureAwait(false);
                }
            }
        }

        public HttpClient GetHttpClient(ITwitterQuery twitterQuery, ITwitterClientHandler twitterHandler = null)
        {
            var handler = (twitterHandler as TwitterClientHandler) ?? new TwitterClientHandler();
            handler.TwitterQuery = twitterQuery;

            var client = new HttpClient(handler)
            {
                Timeout = twitterQuery.Timeout,
            };

            return client;
        }
    }
```

This post is a "duplicate" of this [StackOverflow question](http://stackoverflow.com/questions/36025004/httpclient-does-not-send-same-request-in-w10-uap).

**Comment by linvi** [comment]
Hello David,

I know I just opened this few hours ago, but have you experienced the same problem before.

If yes, do you know some workarounds?
If no, have you been able to identify the problem?

Thanks

**Comment by JonHanna** [comment]
What happens if you explicitly ask for `https://api.twitter.com/1.1/search/tweets.json?q=%D0%9F%D1%80%D0%B8%D0%B2%D0%B5%D1%82!&result_type=mixed&count=100` rather than `https://api.twitter.com/1.1/search/tweets.json?q=%D0%9F%D1%80%D0%B8%D0%B2%D0%B5%D1%82%21&result_type=mixed&count=100`?

**Comment by linvi** [comment]
I cannot test this right now but I am pretty sure that this does work.

But it means that I cannot rely on `Uri.EscapeDataString(queryURL);`.
Is there a "better" method than `Uri.EscapeDataString(queryUrl);` to ensure that `HttpClient` works as expected?

Should I try to escape the data string manually?

**Comment by JonHanna** [comment]
We don't know if those are the right questions until after we see if the explicit uses differ. Unescaping '%21' back to '!' seems like busy-work on first glance since servers should treat them equivalently any time they are used as sub-delim characters (and obviously unescaping in that case would be a clear bug). If the check I suggest does indeed show this to be were the issue is then this would seem to be a bug with twitter's service rather than HttpClient, though that's no comfort to you, and perhaps a common enough bug that HttpClient should still deal with it.

If that does show it to be the case then doing the normal escape and then explicitly unescaping `%21` back to `!` may be a workaround.

**Comment by linvi** [comment]
The problem does not come from Twitter not being able to read the ! or %21.
The problem comes from the fact that as part of Twitter authentication, the OAuthToken needs to be constructed based on the parameters of the queryURL.

When Twitter receives a requests, their api compares the token with the http request url, but because `HttpClient` changed the query parameters, the OAuthToken is no longer valid and Twitter reject the query.

> If that does show it to be the case then doing the normal escape and then explicitly unescaping %21 back to ! may be a workaround.

The real problem comes from the fact that it does not only happen for **!**, it happens for a specific set of characters. And I have no clue of what the complete set of characters is.

**Comment by linvi** [comment]
Just for the sake of sharing, the bug was reported [here](https://github.com/linvi/tweetinvi/issues/120).

And the set of characters the user tested is :

> !
:
(
)
*

**Comment by davidsh** [comment]
@linvi Thx for reporting the bug. As library owner of System.Net.Http, we will investigate. It's possible this problem is not related to HttpClient but rather to System.Uri which is used for parsing the query string parts of the path.

**Comment by davidsh** [comment]
>Hello David,
I know I just opened this few hours ago, but have you experienced the same problem before.
If yes, do you know some workarounds?  If no, have you been able to identify the problem?

Also, please keep in mind that the core team is based in Redmond, WA and we usually are only able to respond to issues during business hours.

**Comment by linvi** [comment]
Hello there, 

I am looking forward to the fix, in the meantime I fixed the problem by creating a new `Uri` from my url and then using the `Uri.AbsoluteUri`.

This value is the same as the value that `HttpClient` is sending.

I hope this can help.

Linvi

**Comment by davidsh** [comment]
@linvi Can you post more of the repro solution for this issue? For example, the project.json and project.lock.json files. That will tell us more of what specific library versions of System.Net.Http you are using.

What specific version of Windows 10 is this?  If you open up a CMD.exe window and type "ver", what is the output?

Regarding the differences you see in the query string parameter escaping:

>On the other side the W10 query parameter finishes by %82%21. Here the question mark has kept its HTML escaping.

Which do you consider correct?  The Console application or the Windows 10 UWP application?  Aren't both the same query string semantically even though the '?' is encoded?

>You can also notice that W10 is adding a cookie.

Where is this cookie coming from? One difference is that in earlier versions of Windows 10 and System.Net.Http, the cookie jar implementation is different.  Can you post the complete Fiddler trace of the Windows 10 UWP app from when it is first installed to when you send this behavior?

**Comment by davidsh** [comment]
@linvi 

Also, in your comment above:
>f you take attention, you will notice that on the console, the end of the query parameter is %82! (with a question mark).
On the other side the W10 query parameter finishes by %82%21. Here the question mark has kept its HTML escaping.

You refer to the '!' character as the "question mark".  But it is actually an "exclamation mark" and not a question mark (which has different encoding).  So, is the character in question `%21`, the '!' character?

**Comment by linvi** [comment]
Repro is as simple as :

``` c#
await client.SendAsync(new HttpRequestMessage(httpMethod, "https://api.twitter.com/1.1/search/tweets.json?q=%D0%9F%D1%80%D0%B8%D0%B2%D0%B5%D1%82%21")).ConfigureAwait(false);
```

Package.json :

``` xml
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="Microsoft.Bcl" version="1.1.10" targetFramework="net45" />
  <package id="Microsoft.Bcl.Async" version="1.0.168" targetFramework="net45" />
  <package id="Microsoft.Bcl.Build" version="1.0.21" targetFramework="net45" />
  <package id="Microsoft.Net.Http" version="2.2.29" targetFramework="net45" />
</packages>
```

`Windows Version` : Microsoft Windows [Version 10.0.10586]

> Which do you consider correct? The Console application or the Windows 10 UWP application? Aren't both the same query string semantically even though the '?' is encoded?

%82 is correct.
%21 should stay %21 and not be changed by the `HttpClient` to `!`.

> Where is this cookie coming from?

I have no idea, I am not setting it.

>Can you post the complete Fiddler trace of the Windows 10 UWP app from when it is first installed to when you send this behavior?

What I posted above are the only 2 traces that I can give you as they are always the same.

> You refer to the '!' character as the "question mark". But it is actually an "exclamation mark" and not a question mark (which has different encoding). So, is the character in question %21, the '!' character?

The `%21` is an exclamation mark. Sorry for the confusion.

**Comment by davidsh** [comment]
Thx for the added information.

The packages you listed are not for a Windows 10 UWP.

For example, "Microsoft.Net.Http" is only for a Windows 8.x application:
http://www.nuget.org/packages/Microsoft.Net.Http/2.2.29

So, perhaps you are not building a Windows 10 UWP but actually a Windows 8.x Store Application but running it on Windows 10?   Windows 10 UWP application projects have a 'project.json' file and not a "packages.json" file.   And for HttpClient, the package is usually 'System.Net.Http':

http://www.nuget.org/packages/System.Net.Http

So, I'm confused as to what kind of application you are actually building.

Does your application fail in some way?  I see that the '!' character is percent-encoded in one application but decoded again by the Store App you wrote.  But is the application failing?  Or is this difference you noted not really affecting the application, but is just a difference you noted when looking at the wire traces?

**Comment by linvi** [comment]
The code I am working on is a Portable Class Library compatible with W10 UAP.
The library works for every target (windows, linux, wphone, iOS...).

But as I mentioned, the `HttpClient` targeting the W10 platform is acting differently than on the other platforms.

So my W10 UAP project.json is the following :

``` json
{
  "dependencies": {
    "Autofac": "3.5.2",
    "Microsoft.ApplicationInsights": "1.2.3",
    "Microsoft.ApplicationInsights.PersistenceChannel": "1.2.3",
    "Microsoft.ApplicationInsights.WindowsApps": "1.1.1",
    "Microsoft.Bcl": "1.1.10",
    "Microsoft.Bcl.Async": "1.0.168",
    "Microsoft.Bcl.Build": "1.0.21",
    "Microsoft.Net.Http": "2.2.29",
    "Microsoft.NETCore.UniversalWindowsPlatform": "5.0.0",
    "Newtonsoft.Json": "8.0.2"
  },
  "frameworks": {
    "uap10.0": {}
  },
  "runtimes": {
    "win10-arm": {},
    "win10-arm-aot": {},
    "win10-x86": {},
    "win10-x86-aot": {},
    "win10-x64": {},
    "win10-x64-aot": {}
  }
}
```

> Does your application fail in some way? I see that the '!' character is percent-encoded in one application but decoded again by the Store App you wrote. But is the application failing? Or is this difference you noted not really affecting the application, but is just a difference you noted when looking at the wire traces?

Well I have been able to find a workaround for this issue as mentioned [here](https://github.com/dotnet/corefx/issues/6946#issuecomment-198814619) but it is a problem.

The reason why this is a problem is the fact that `HttpClient` is not executing the expected query. In the case of Twitter, the authentication mechanism creates an AuthenticationToken that is based on the user credentials AND on the parameters of the query.
Therefore, when I create the AuthenticationToken, it is based on the `%21` version of the query. But as `HttpClient` executes the `!` version of the query, Twitter detects that the Token is not associated with the query `HttpClient` attempts to execute and therefore refuses the access.

I understand this is a problem that could be quite rare, but I do not think it is normal for `HttpClient` to not do exactly what the developer ask him to do.

**Comment by davidsh** [comment]
I have root caused this issue. It is actually a bug in the .NET Framework System.Uri class. This bug would also be present in the CoreFx System.Uri class (since it was ported from .NET Framework Desktop) and hence would show up in CoreFx HTTP stack. But the bug will not occur in the UWP HTTP stack.

Given 2 Uri's:

```c#
var CustomUri1 = new Uri("http://www.example.com?q=\u041f\u0440\u0438\u0432\u0435\u0442%21");
var CustomUri2 = new Uri("http://www.example.com?q=abcdef%21");
```

On .NET Framework or CoreFx (normal console application):
#1 output (note the %21 gets changed to "!"
>Uri.OriginalString=http://www.example.com?q=Привет%21
Uri.AbsoluteUri=http://www.example.com/?q=%D0%9F%D1%80%D0%B8%D0%B2%D0%B5%D1%82!

#2 output (the %21 stays as-is)
>Uri.OriginalString=http://www.example.com?q=abcdef%21
Uri.AbsoluteUri=http://www.example.com/?q=abcdef%21

On UWP, the HTTP stack will see this output:
#1 output (the %21 stays as-is)
>Uri.OriginalString=http://www.example.com?q=Привет%21
Uri.AbsoluteUri=http://www.example.com/?q=%D0%9F%D1%80%D0%B8%D0%B2%D0%B5%D1%82%21

#2 output (the %21 stays as-is)
>Uri.OriginalString=http://www.example.com?q=abcdef%21
Uri.AbsoluteUri=http://www.example.com/?q=abcdef%21

This is not a bug in the HTTP stack, per se, but rather in the System.Uri class in the .NET Framework. For UWP, the HTTP stack calls into WinRT APIs (Windows.Web.Http). And by calling into WinRT APIs, the System.Uri is converted into a WinRT Windows.Foundation.Uri class. The conversion process uses the .OriginalString of the System.Uri. This helps to preserve the %21. In addition, the WinRT Uri class doesn't have this bug where the %21 sometimes (not always) gets converted back to the "!" character.

In summary, the System.Uri class has a bug where the %21 is being decoded back to "!" at times. It seems to stay as %21 if other non-encoded characters (such as "abcdef" are present). But if encoded characters from the Unicode code set are present, then the %21 seems to revert to "!" character.

**Comment by linvi** [comment]
Hi,

Thank you for all this precision, I noticed the problem with Uri. But I did not go as far as you did.
Do I need to create a bug report somewhere else?

Cheers,
Linvi

**Comment by davidsh** [comment]
No. We will create additional bugs in other databases as needed and will update this bug accordingly.

**Comment by jofriedm-msft** [comment]
cc @mirobers

We are observing a similar problem with '[' and ']' characters where we need to add an authentication header that is based on the escaped URL. We have been relying on the on the uri.AbsolutePath to get the escaped string that is sent on the wire in order for signing to match what the server sees. 

Will this fix also resolve our situation? Is there an ETA for this?

**Comment by davidsh** [comment]
cc: @jofriedm-msft @mirobers

>We are observing a similar problem with '[' and ']' characters where we need to add an authentication header that is based on the escaped URL. We have been relying on the on the uri.AbsolutePath to get the escaped string that is sent on the wire in order for signing to match what the server sees.

It turns out your problem is different from this issue.  I opened a new issue #11156.

**Comment by karelz** [comment]
This is side-effect of the fact we have different HTTP stacks in Desktop, Core and UWP.
Fixing this specific bug will be very complex and difficult. It would be better to focus on reconciling and merging the Http stacks implementations.

If you think this bug has high impact on customers, please let us know.
